VIB-68: [Sub-Epic] Make Changes to Database (part of Timesheet Entry Feature epic)

PivtoranisV opened the pull request:
[![VIB-68](https://badgen.net/badge/JIRA/VIB-68/009900)](https://cloverpop-internship-2025.atlassian.net/browse/VIB-68) [<img width="16" alt="Powered by Pull Request Badge" src="https://user-images.githubusercontent.com/1393946/111216524-d2bb8e00-85d4-11eb-821b-ed4c00989c02.png">](https://pullrequestbadge.com/?utm_medium=github&utm_source=wahanegi&utm_campaign=badge_info)<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->

valvolyk commented on app/models/project.rb:
```diff
@@ -0,0 +1,56 @@
+  scope :active, -> { where(deleted_at: nil) }
+  scope :deleted, -> { where.not(deleted_at: nil) }
+
+  def soft_delete!
```
**Optional**:
It's a big chance that we will use this func-ty somewhere else) So to make this functionality reusable across multiple models, you can create a module, often referred to as a concern, that encapsulates the soft delete logic. 

```
module SoftDeletable
  extend ActiveSupport::Concern

  included do
    scope :deleted, -> { where.not(deleted_at: nil) }
    scope :not_deleted, -> { where(deleted_at: nil) }
  end

  def soft_delete!
    update(deleted_at: Time.current)
  end

  def restore!
    update(deleted_at: nil)
  end

  def deleted?
    deleted_at.present?
  end
end
```

```
class YourModel < ApplicationRecord
  include SoftDeletable
end
```

PivtoranisV commented on app/admin/projects.rb:
```diff
@@ -0,0 +1,72 @@
+      else
+        project.destroy!
+        redirect_to admin_projects_path, notice: "Project was permanently deleted!"
+      end
```
Hi, @rogergraves and @valvolyk, thank you for the comments, I learned quite a lot by making a SoftDeletable module and overriding the destroy action. And now we do not repeat code around the project model

I have overridden only destroy action, not destroy! (with a bang), and make sure we use only destroy action in our project (I'm not 100% sure if it is the right way to do it). 

Also, I did a quick test, and it looks like it works as expected, you can see an example here:
https://www.loom.com/share/66d7a6ff31fc4daf9629238473e70d33?sid=2411cc92-2f5f-42f4-97ba-4a3a3687bf21

rogergraves reviewed (commented):
Excellent work @PivtoranisV!
In the Loom video you mentioned the difference between `project.destroy` (which soft-deletes if there are associated timesheets) and `project.destroy!` which should kill the record regardless. This is a great approach. I would suggest that `project.destroy!` should also delete orphaned `time_sheet_entries`. If you agree:
 * Add a `dependent: :destroy` to the project's `has_many :time_sheet_entries`
 * Add some test coverage that shows `project.destroy` versus `project.destroy!` when there are children project.time_sheet_entries` present.

PivtoranisV commented:
> Excellent work @PivtoranisV! In the Loom video you mentioned the difference between `project.destroy` (which soft-deletes if there are associated timesheets) and `project.destroy!` which should kill the record regardless. This is a great approach. I would suggest that `project.destroy!` should also delete orphaned `time_sheet_entries`. If you agree:
> 
> * Add a `dependent: :destroy` to the project's `has_many :time_sheet_entries`
> * Add some test coverage that shows `project.destroy` versus `project.destroy!` when there are children project.time_sheet_entries` present.

Hi @rogergraves, I added `dependent: :destroy` and updated the test, which includes testing of `associations`, `soft_delete!`, `restore!`, `destroy` and `destroy!` methods

rogergraves commented on spec/models/project_spec.rb:
```diff
@@ -94,22 +98,56 @@
+        end
+      end
+
+      context 'with time_sheet_entries' do
+        it 'performs a soft deletion and leaves associated time_sheet_entries intact' do
+          expect { project_with_entries.destroy! }.not_to change(Project, :count)
+          expect { project_with_entries.destroy! }.not_to change(TimeSheetEntry, :count)
+
+          expect(project_with_entries.reload.deleted_at).to be_present
+          expect(TimeSheetEntry.find_by(id: time_entry.id)).to be_present
+        end
```
@PivtoranisV I was thinking that `destroy!` should kill the record and its children, not do the same thing as `destroy`, which is to only delete if there are no children and soft_delete if there are cnildren. The use case for `destroy!` would be to remove test data (in console).

PivtoranisV replied:
Thank you, @rogergraves, for another great comment so I can learn something new. 

To achieve this behaviour where we can use `destroy!` for scenarios like cleaning up test data in the Rails console, I had to override `destroy!` method in project model, as the default behaviour was soft deleting projects with associated entries. By overriding it, I’m now explicitly calling destroy_all on the associated entries before invoking super to perform the hard delete.

I’ve also updated the test cases accordingly to reflect this change. Also, I have tested locally in active admin, and all works as expected

Please let me know if this approach is correct.

rogergraves commented on spec/models/project_spec.rb:
```diff
@@ -94,22 +98,56 @@
   context 'Soft delete functionality' do
-    it 'soft deletes a project by setting deleted_at' do
-      expect { project.soft_delete! }.to change { project.reload.deleted_at }.from(nil)
+    describe '#soft_delete!' do
```
It's not clear to me what the difference is between `soft_delete` and `soft_delete!`, also `restore` versus `restore!`. If there is no difference then we shouldn't use these `!` variants IMO.

PivtoranisV replied:
Thank you, Roger! I have updated the code, and now we are using the `soft_delete` and `restore` methods.

rogergraves reviewed (approved):
👍🏻

Maryna-Harasko commented:
Will be retested in the VIB-45 branch
